### DISCUSSION TITLE
What would be the cleaner way of picking a Postgres Schema with the Async API?

### DISCUSSION THREAD

**Discussion by cblegare** [sqlalchemy/sqlalchemy#6898] [Usage Questions] [answered]
Hello there!

I'd like to connect to a non-default postgres schema with the the asyncpg dialect.  It is quite easy with the default sync dialect:

```python
create_engine(
    uri,
    connect_args={
        "options": "-csearch_path={}".format("myschema")
    },
)
```

and I'd like a similar way of picking a schema with the Async driver,  I have found this comment that makes me think that should be possible

> To clarify, search_path=foo would work in the DSN too.
> -- https://github.com/MagicStack/asyncpg/issues/438#issuecomment-734510107

But I couldn't find the proper way to configure the dialect using `create_async_engine` that way.

Any clue?

**Comment by zzzeek** [comment]
you should use the "connect" event, against the "sync" engine, to set up search path as you want right up front:

```py
import asyncio

from sqlalchemy import event
from sqlalchemy import text
from sqlalchemy.ext.asyncio import create_async_engine


e = create_async_engine("postgresql+asyncpg://scott:tiger@localhost/test")

@event.listens_for(e.sync_engine, "connect")
def set_search_path(dbapi_connection, connection_record):
    cursor = dbapi_connection.cursor()
    cursor.execute("SET search_path TO test_schema")
    cursor.close()


async def main():
    async with e.connect() as conn:
        x = await conn.scalar(text("SHOW search_path"))
        print(x)

asyncio.run(main())
```

**Reply by cblegare** [reply]
> `@event.listens_for(e.sync_engine, "connect")`

This is the part I missed! I tried with `@event.listens_for(e, "connect")` to no aval

Thank you kind sir, you are a gentleman and a scholar!

**Reply by CaselIT** [reply]
looking at https://magicstack.github.io/asyncpg/current/api/index.html I think you could also specify it as 
```py
create_async_engine(
    uri,
    connect_args={
        "server_settings": {"search_path":"my_schema"}
    },
)
```

**Reply by zzzeek** [reply]
yes I have been lazy and not documented how to do event listening in the asyncio section, we really need that

**Reply by cblegare** [reply]
@CaselIT 

~~I do not have the code at hand, but I believe to have tried this one without success.  I'm sorry I can't confirm nor infirm at the moment.~~

It does work. I even prefer this one, even though it is driver-specific.  Thank you!

**Reply by CaselIT** [reply]
Great!

**Reply by cblegare** [reply]
@zzzeek 
Looking at the very good documentation of SQLAlchemy, "lazy" is not the word that comes to my mind.  Thank you all for this amazing piece of work

**Reply by CaselIT** [reply]
> 
> 
> yes I have been lazy and not documented how to do event listening in the asyncio section, we really need that

I've added https://github.com/sqlalchemy/sqlalchemy/issues/6899 so we better remember. We have had a lot of questions regarding it

**Comment by CaselIT** [comment, marked as answer]
looking at https://magicstack.github.io/asyncpg/current/api/index.html I think you could also specify it as
```py
create_async_engine(
    uri,
    connect_args={
        "server_settings": {"search_path": "my_schema"}
    },
)
```

copied from https://github.com/sqlalchemy/sqlalchemy/discussions/6898#discussioncomment-1201831

**Reply by Kokleng-Dev** [reply]
Hello sir, it work for me. thank you.
but i have one question, if i have multi schema, so i need to create more create_async_engine for difference schema?

**Reply by CaselIT** [reply]
depends what you want to do, if you are planning on using `"server_settings": {"search_path": "my_schema"}` then yes.

as an alternative there is the schema translate map feature of sqlalchemy: https://docs.sqlalchemy.org/en/20/core/connections.html#schema-translating
When you use it you can have multiple "virtual" engines connected to different translate maps by using https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Engine.execution_options , something like
```py
_base_engine = create_async_engine(...)
schema1_engine = _base_engine.execution_options(...)
schema2_engine = _base_engine.execution_options(...)
```

The main difference is that while the first approach will create independent engines, with different connection pools, in the second approach the all engines share the same pool
